src/dune_pkg/rev_store.ml

```diff
@@ -661,205 +726,81 @@ module At_rev = struct
-    default_branch
-  ;;
-
-  let rev_of_name { repo; handle; source; default_branch = _; add_remote } ~name =
-    (* TODO handle non-existing name *)
-    let* rev = run_capture_line repo [ "rev-parse"; sprintf "%s/%s" handle name ] in
-    let revision = Rev.Rev rev in
-    At_rev.of_rev repo ~add_remote ~revision ~source >>| Option.some
-  ;;
-
-  let rev_of_ref { repo; handle = _; source; default_branch = _; add_remote } ~ref =
-    let revision = Rev.Rev ref in
-    At_rev.of_rev repo ~add_remote ~revision ~source >>| Option.some
-  ;;
-end
-
-let remote_exists dir ~name =
-  let stderr_to = make_stderr () in
-  let command = [ "remote"; "--verbose" ] in
-  let+ lines =
-    let git = Lazy.force Vcs.git in
-    Process.run_capture_lines ~dir ~display:Quiet ~stderr_to ~env Strict git command
-  in
-  lines
-  |> List.find ~f:(fun line ->
-    match String.lsplit2 ~on:'\t' line with
-    | None -> false
-    | Some (candidate, _) -> String.equal candidate name)
-  |> Option.is_some
-;;
-
-let query_head_branch =
+let remote =
+  let hash = Re.(rep1 alnum) in
+  let head_mark, head = Re.mark (Re.str "HEAD") in
   let re =
     Re.(
       compile
       @@ seq
            [ bol
-           ; str "ref: refs/heads/"
-           ; group (rep1 (diff any space))
+           ; group hash
            ; rep1 space
-           ; str "HEAD"
-           ; eol
+           ; alt
+               [ head
+               ; seq
+                   [ str "refs/"
+                   ; alt [ str "heads"; str "tags" ]
+                   ; str "/"
+                   ; group (rep1 any)
+                   ]
+               ]
            ])
   in
-  fun t source ->
-    let+ lines =
-      run_capture_lines
-        t
-        ~display:!Dune_engine.Clflags.display
-        [ "ls-remote"; "--symref"; source ]
-    in
-    List.find_map lines ~f:(fun line ->
-      match Re.exec_opt re line with
-      | None -> None
-      | Some m -> Some (Re.Group.get m 1))
-;;
-
-let branch_of_refspec refspec =
-  refspec
-  |> String.drop_prefix_if_exists ~prefix:"+"
-  |> String.lsplit2 ~on:':'
-  |> Option.bind ~f:(fun (remote_ref, _local_ref) ->
-    String.drop_prefix remote_ref ~prefix:"refs/heads/")
-;;
-
-let find_section =
-  let re =
-    Re.seq
-      [ Re.(rep space)
-      ; Re.str "[remote "
-      ; Re.char '"'
-      ; Re.group (Re.rep (Re.diff Re.any (Re.char '"')))
-      ; Re.char '"'
-      ; Re.char ']'
-      ; Re.(rep space)
-      ]
-    |> Re.compile
-  in
-  fun contents ~name ->
-    let rec loop (xs : Re.split_token list) =
-      match xs with
-      | [] -> None
-      | `Text _ :: rest -> loop rest
-      | `Delim delim :: rest ->
-        if Re.Group.get delim 1 = name
-        then
-          Some
-            (match rest with
-             | `Text s :: _ -> s
-             | _ -> "")
-        else loop rest
+  fun t ~url ->
+    let f url =
+      let command = [ "ls-remote"; url ] in
+      let refs =
+        Fiber_lazy.create (fun () ->
```

Comment:
Why is this a `Fiber_lazy` if gets forced right after? Am I missing something here?

Reply:
It gets forced only when we bind on the fiber that forces it. We don't bind on this fiber in this function